### ISSUE TITLE
Unable to upload >5mb files with Crystal 1.0.0

### ISSUE THREAD

**Issue by OldhamMade** [taylorfinnell/awscr-s3#97]
Using a slightly modified version of `examples/file_upload.cr` on Crystal 1.0.0, and uploading a 10mb image file, I'm seeing the following error:

```
Unhandled exception in spawn: The Content-MD5 you specified did not match what we received. (Awscr::S3::BadDigest)
  from lib/awscr-s3/src/awscr-s3/http.cr:116:50 in 'handle_response!'
  from lib/awscr-s3/src/awscr-s3/http.cr:88:16 in 'exec'
  from lib/awscr-s3/src/awscr-s3/http.cr:46:7 in 'put'
  from lib/awscr-s3/src/awscr-s3/client.cr:111:14 in 'upload_part'
  from lib/awscr-s3/src/awscr-s3/multipart_file_uploader.cr:87:17 in 'upload_part'
  from lib/awscr-s3/src/awscr-s3/multipart_file_uploader.cr:80:9 in '->'
  from /opt/crystal/1.0.0/src/primitives.cr:255:3 in 'run'
  from /opt/crystal/1.0.0/src/fiber.cr:92:34 in '->'
```

The first 5.2mb is uploaded, but then the above error is raised and the method returns `true`. This is with an S3-compatible service.

If I try to use Amazon's S3 service, the error is as follows:

```
Unhandled exception in spawn: XAmzContentSHA256Mismatch: The provided 'x-amz-content-sha256' header does not match what was computed. (Awscr::S3::Exception)
  from lib/awscr-s3/src/awscr-s3/http.cr:116:50 in 'handle_response!'
  from lib/awscr-s3/src/awscr-s3/http.cr:88:16 in 'exec'
  from lib/awscr-s3/src/awscr-s3/http.cr:46:7 in 'put'
  from lib/awscr-s3/src/awscr-s3/client.cr:111:14 in 'upload_part'
  from lib/awscr-s3/src/awscr-s3/multipart_file_uploader.cr:87:17 in 'upload_part'
  from lib/awscr-s3/src/awscr-s3/multipart_file_uploader.cr:80:9 in '->'
  from /opt/crystal/1.0.0/src/primitives.cr:255:3 in 'run'
  from /opt/crystal/1.0.0/src/fiber.cr:92:34 in '->'
```

Either way – the "large" file does not upload successfully. 

Using a 1.1mb image file, uploads are successful as expected, for both services.

**Comment by snadon** [comment]
Related issue: #96

**Comment by taylorfinnell** [comment]
Guessing here, I will try to look soon. But I suspect it's an issue calculating the part sizes. Anything above 5MB uses multi part uploading.

**Comment by taylorfinnell** [comment]
Ok, I replicated a file close to 5MB

```
dd if=/dev/urandom of=test.img bs=1 count=0 seek=5500000
```

If the `spawn` and channel code is removed it uploads fine. So it's something to do with that

**Comment by taylorfinnell** [comment]
Also, when the spawns _are_ in the code. I get OpenSSL errors every 3rd run or so

**Comment by OldhamMade** [comment]
> Also, when the spawns _are_ in the code. I get OpenSSL errors every 3rd run or so. 

I too am seeing this error. I thought it was a problem with my setup. Waiting a little while between tests seemed to work around this problem during testing, though would be good to have this resolved too if it persists after the above changes.

**Comment by taylorfinnell** [comment]
The issue goes away with this single change

```
diff --git a/src/awscr-s3/http.cr b/src/awscr-s3/http.cr
index 554bf60..102ef46 100644
--- a/src/awscr-s3/http.cr
+++ b/src/awscr-s3/http.cr
@@ -136,7 +136,7 @@ module Awscr::S3

     # :nodoc:
     private def http
-      @client ||= create_client(@endpoint, @signer)
+      @client = create_client(@endpoint, @signer)
     end

     private def create_client(endpoint, signer)
```

But, we do want to be able to reuse the client, so this isn't the fix - but a clue

**Comment by confact** [comment]
How is this going, @taylorfinnell? Do you have a branch with that small fix I can use temporarily? I would love to get this fixed. So please tell me if I can do anything to get this move forward.

**Comment by taylorfinnell** [comment]
I don’t - I still don’t know if this is the real fix. You could monkey patch the method in your project to see if it helps you

**Comment by taylorfinnell** [comment]
This is probably fixed by #103 if someone wants to give it a shot

**Comment by stakach** [comment]
This is exactly the issue I was solving in #103. Didn't look at existing issues though
